Stop using assetstore; don't download already-annexed blobs; stop downloading at assets without SHA256

jwodder opened the pull request:
Closes #84.

jwodder committed:
Some code refactoring

jwodder committed:
Stop using assetstore; don't download already-annexed blobs; stop downloading at assets without SHA256

jwodder committed:
Fully handle not-yet-downloaded assets

jwodder commented:
The tests are failing because of [this issue](https://github.com/dandi/dandi-api/issues/548).

yarikoptic commented on tools/backups2datalad.py:
```diff
@@ -191,213 +185,21 @@ def sync_dataset(self, dandiset: RemoteDandiset, ds: Dataset) -> bool:
+                syncer.dump_asset_metadata()
+            if any(r["state"] != "clean" for r in ds.status()):
+                log.info("Commiting changes")
+                with custom_commit_date(dandiset.version.modified):
```
just a note: with all the "metadata mutations" this might differ from the `latest([a.modified for a in all_assets])` but I think it is ok.  The main point is to have a date reflective of changes, and also this one is likely to relate to dandiset metadata modification date

yarikoptic commented on tools/backups2datalad.py:
```diff
@@ -817,10 +854,12 @@ def sanitize_contentUrls(urls: List[str]) -> List[str]:
-    for f in ("dandiset_id", "version_id"):
-        adict.pop(f, None)
+    # due to  https://github.com/dandi/dandi-api/issues/231
+    # we need to sanitize temporary URLs. TODO: remove when "fixed"
```
hm, that https://github.com/dandi/dandi-api/issues/231 is closed in July -- if that is still the issue, could you please reopen and provide further detail

jwodder replied:
It no longer seems to be a problem.

jwodder committed:
Increase Dandiset validity wait time in tests

According to [1], dandi-staging runs validations every 60 seconds, so let's try
waiting for 65.

[1]: https://github.com/dandi/dandi-api/issues/548#issuecomment-934616801

yarikoptic commented on tools/backups2datalad.req.txt:
```diff
@@ -1,8 +1,11 @@
 click-loglevel ~= 0.2
-dandi >= 0.27.0
+#dandi >= 0.27.0
+dandi @ git+https://github.com/dandi/dandi-cli.git@asset-created
```
```suggestion
dandi >= 0.29.0
```

jwodder replied:
Applied independently.

yarikoptic reviewed (commented):
forgot to "commit" some minor comments I have left.

Great to see that CI is happy with extended (`65`) wait. Could you please test it "in real life" on 000029 ?

jwodder committed:
Update dandi requirement

jwodder committed:
Remove dandi/dandi-api#231 workaround

jwodder commented:
@yarikoptic I ran the script (almost) successfully on 000029.  First I ran it against a local clone of the dataset, and it updated some metadata; then I ran it on an empty directory, and it crashed [on this line](https://github.com/dandi/dandisets/blob/84e332a0764329982726410356d9176e15fdb4b2/tools/backups2datalad.py#L343) when it tried to determine where to tag the published versions despite there only being one backups2datalad commit.

yarikoptic commented:
well, we should never run on a completely empty repo -- `datalad create` should already initiate some commits... could you please check if we are using the `created` of the dandiset for that commit time stamp?

overall then -- great if works!  I guess I would just wait now for an ongoing run to finish up.  Meanwhile since this PR doesn't take care about #89 or #90 (if I got it right) -- could you approach them on top of this branch?

jwodder commented:
@yarikoptic The script did make some commits, but the "identify commit to tag" code only considers commits with "[backups2datalad]" in the message, which there was only one of, and this commit's timestamp equalled the modification timestamp of the "draft" version, which is later than the created timestamps of all of the published versions, so [this block](https://github.com/dandi/dandisets/blob/84e332a0764329982726410356d9176e15fdb4b2/tools/backups2datalad.py#L340-L342) failed to add anything to `candidates`, resulting in a failed assertion.

yarikoptic commented:
I have ideas on how such case could happen in real life, but I guess it shouldn't happen.  We have an assertion so if we run into it -- we will react.

jwodder committed:
Remove some no-longer-needed code

jwodder committed:
Make nox run "typing" first and exit if it fails

jwodder committed:
Mark the first digestless asset as "future"
